Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,560 @@
+capture type is an `Alternation` that's generic over each option's underlying
+pattern.
+
+```swift
+/([01]+)|[0-9]+|([0-9A-F]+)/
+// => `Regex<Alternation<(Substring, Void, Substring)>>`
+```
+
+If there are no capturing groups within an alternation the resulting capture type is `Void`.
+
+```swift
+/[01]+|[0-9]+|[0-9A-F]+/
+// => `Regex<Void>`
+```
+
+Nested captures follow the algebra previously described.
+
+```swift
+/([01]+|[0-9]+|[0-9A-F]+)/
+// => `Regex<Substring>`
+/(([01]+)|([0-9]+)|([0-9A-F]+))/
+// => Regex<Substring, Alternation<(Substring, Substring, Substring)>>
+```
+
+At the use site, you often want to be able use `Alternation` like an `enum`.
+Ideally you'd be able to exhaustively switch over all the captures.
+
+Like:
+
+```swift
+let number = line
+    .firstMatch(/([01]+)|([0-9]+)|([0-9A-F]+)/)?
+    .captures
+    .flatMap {
+        switch $0 {
+        case let .0(binary):
+            return Int(binary, radix: 2)
+        case let .1(decimal):
+            return Int(decimal, radix: 10)
+        case let .2(hex):
+            return Int(decimal, radix: 16)
+        }
+    }
+```
+
+Or:
+
+```swift
+let number = line
+    .firstMatch(/(?<binary>[01]+)|(?<decimal>[0-9]+)|(?<hex>[0-9A-F]+)/)?
+    .captures
+    .flatMap {
+        switch $0 {
+        case let .binary(str):
+            return Int(str, radix: 2)
+        case let .decimal(str):
+            return Int(str, radix: 10)
+        case let .hex(str):
+            return Int(str, radix: 16)
+        }
+    }
+```
+
+In the fullness of time, we'd like to design a language feature to support this.
+In the meantime, we'd like to do the best we can and leave the door open for a
+source compatible migration.
+
+With variadic generics we think we can defining the following `Alternation` type.
+
+```swift
+struct Alternation<Captures> { ... }
+
+extension<T...> Alternation where Captures == (T...) {
+    var captures: (T?...) { ... }
+}
+```
+
+Which would support:
+
+```swift
+let number = line
+    .firstMatch(/([01]+)|([0-9]+)|([0-9A-F]+)/)?
+    .captures
+    .flatMap {
+        switch $0.captures {
+        case let (binary?, nil, nil):
+            return Int(binary, radix: 2)
+        case let (nil, decimal?, nil):
+            return Int(decimal, radix: 10)
+        case let (nil, nil, hex?):
+            return Int(hex, radix: 16)
+        default:
+            fatalError("unreachable")
+        }
+    }
+```
+
+Or:
+
+```swift
+let number = line
+    .firstMatch(/(?<binary>[01]+)|(?<decimal>[0-9]+)|(?<hex>[0-9A-F]+)/)?
+    .captures
+    .flatMap {
+        if let binary = $0.captures.binary {
+            return Int(binary, radix: 2)
+        } else if let decimal = $0.captures.decimal {
+            return Int(decimal, radix: 10)
+        } else if let hex = $0.captures.hex {
+            return Int(hex, radix: 16)
+        } else {
+            fatalError("unreachable")
+        }
```

Comment:
We could eliminate some this `.captures` boilerplate by using `@dynamicMemberLookup` on the `Alternation` type:

```swift
if let binary = $0.binary {
    return Int(binary, radix: 2)
} else if let decimal = $0.decimal {
    return Int(decimal, radix: 10)
} else if let hex = $0.hex {
    return Int(hex, radix: 16)
} else {
    fatalError("unreachable")
}
```

Reply:
Added!